docs/docs/dev_docs/contracts/portals/main.md

```diff
@@ -0,0 +1,331 @@
+:::
+
+As earlier, we can use a token bridge as an example. In this case, we are burning tokens on L2 and sending a message to the portal to free them on L1.
+
+```rust title="NonNativeTokenContract.nr"
+// burns token and sends a message to the portal
+fn withdraw(
+    amount: Field,
+    sender: Field,
+    recipient: Field,
+    callerOnL1: Field,
+) -> distinct pub abi::PrivateCircuitPublicInputs {
+    ...
+    let sender_balance = storage.balances.at(sender);
+    spend_notes(&mut context, sender_balance, amount, sender);
+
+    let content = get_withdraw_content_hash(amount, recipient, callerOnL1);
+    context.message_portal(content);
+    ...
+}
+```
+
+When the transaction is included in a rollup block it will be inserted into the `Outbox`, where the recipient portal can consume it from. When consuming, the `msg.sender` must match the `recipient` meaning that only portal can actually consume the message.
+
+```solidity title="IOutbox.sol"
+struct L2ToL1Msg {
+    DataStructures.L2Actor sender;
+    DataStructures.L1Actor recipient;
+    bytes32 content;
+}
+
+function consume(DataStructures.L2ToL1Msg memory _message) 
+    external 
+    returns (bytes32 entryKey);
+```
+
+As noted earlier, the portal contract should check that the sender is as expected. In the example below, we support only one sender contract (stored in `l2TokenAddress`) so we can just pass it as the sender, that way we will only be able to consume messages from that contract. If multiple senders are supported, you could use a have `mapping(address => bool) allowed` and check that `allowed[msg.sender]` is `true`.
+
+```solidity title="TokenPortal.sol"
+function withdraw(uint256 _amount, address _recipient, bool _withCaller)
+    external
+    returns (bytes32)
+  {
+    // Create the message structure
+    DataStructures.L2ToL1Msg memory message = DataStructures.L2ToL1Msg({
+      sender: DataStructures.L2Actor(l2TokenAddress, version),
+      recipient: DataStructures.L1Actor(address(this), block.chainid),
+      content: Hash.sha256ToField(
+        abi.encodeWithSignature(
+          "withdraw(uint256,address,address)",
+          _amount,
+          _recipient,
+          _withCaller ? msg.sender : address(0)
+        )
+        )
+    });
+
+    // Consume the message
+    bytes32 entryKey = registry.getOutbox().consume(message);
+
+    // Transfer the tokens to the user
+    underlying.transfer(_recipient, _amount);
+
+    return entryKey;
+  }
+```
+
+
+## How to deploy a contract with a portal
+- Deploy L1 
+- Deploy l2
+- Initialize l1 with l2 address.
+
+## Standards
+
+### Structure of messages
+The application developer should consider creating messages that follow a structure like function calls, e.g., using a function signature and arguments. This will make it easier to ensure that they are not making messages that could be misinterpreted by the recipient. 
```

Comment:
```suggestion
The application developer should consider creating messages that follow a function call structure e.g., using a function signature and arguments. This will make it easier to prevent producing messages that could be misinterpreted by the recipient. 
```